File: exercise_0/solution.test.ts
import { sumList, boundedSum } from "./solution";

const MAX_LEVEL = process.env.LEVEL ? parseInt(process.env.LEVEL) : Infinity;
const level = (n: number, name: string, fn: () => void) =>
  (n <= MAX_LEVEL ? describe : describe.skip)(`Level ${n}: ${name}`, fn);

// ── Level 1: Basic Sum ────────────────────────────────────────────────────────

level(1, "sumList", () => {
  test("empty list returns 0", () => {
    expect(sumList([])).toBe(0);
  });

  test("single element", () => {
    expect(sumList([42])).toBe(42);
  });

  test("positive numbers", () => {
    expect(sumList([1, 2, 3])).toBe(6);
  });

  test("negative numbers", () => {
    expect(sumList([-1, -2, -3])).toBe(-6);
  });

  test("mixed numbers", () => {
    expect(sumList([-1, -2, 5])).toBe(2);
  });

  test("1 to 100 sums to 5050", () => {
    const numbers = Array.from({ length: 100 }, (_, i) => i + 1);
    expect(sumList(numbers)).toBe(5050);
  });

  test("all zeros", () => {
    expect(sumList([0, 0, 0])).toBe(0);
  });

  test("single negative", () => {
    expect(sumList([-7])).toBe(-7);
  });
});

// ── Level 2: Bounded Sum ──────────────────────────────────────────────────────

level(2, "boundedSum", () => {
  test("all in range", () => {
    expect(boundedSum([1, 2, 3], 1, 3)).toBe(6);
  });

  test("some in range", () => {
    expect(boundedSum([1, 2, 3, 4, 5], 2, 4)).toBe(9);
  });

  test("none in range returns 0", () => {
    expect(boundedSum([1, 2, 3], 10, 20)).toBe(0);
  });

  test("negative range", () => {
    expect(boundedSum([-5, -1, 0, 3], -1, 3)).toBe(2);
  });

  test("exact boundary match", () => {
    expect(boundedSum([7], 7, 7)).toBe(7);
  });

  test("empty list returns 0", () => {
    expect(boundedSum([], 0, 10)).toBe(0);
  });

  test("inclusive lower boundary", () => {
    expect(boundedSum([1, 2, 3, 4], 2, 10)).toBe(9);
  });

  test("inclusive upper boundary", () => {
    expect(boundedSum([1, 2, 3, 4], 0, 3)).toBe(6);
  });

  test("single element out of range", () => {
    expect(boundedSum([5], 6, 10)).toBe(0);
  });

  test("all same values in range", () => {
    expect(boundedSum([3, 3, 3, 3], 3, 3)).toBe(12);
  });
});

File: exercise_0/SolutionTest.java
import java.util.*;

public class SolutionTest {
  static final List<Row> rows = new ArrayList<>();
  static int maxLevel = Integer.MAX_VALUE;
  interface Fn { void run() throws Exception; }
  record Row(String name, String status, String error, int line) {}
  public static void main(String[] args) {
    if (args.length > 0) maxLevel = Integer.parseInt(args[0]);
    test(1, "empty list returns 0", 10, () -> eq(0, Solution.sumList(new int[] {})));
    test(1, "positive numbers", 11, () -> eq(6, Solution.sumList(new int[] {1, 2, 3})));
    test(1, "mixed numbers", 12, () -> eq(2, Solution.sumList(new int[] {-1, -2, 5})));
    test(1, "1 to 100 sums to 5050", 13, () -> {
      int[] values = new int[100];
      for (int i = 0; i < values.length; i++) values[i] = i + 1;
      eq(5050, Solution.sumList(values));
    });
    test(2, "bounded sum all in range", 18, () -> eq(6, Solution.boundedSum(new int[] {1, 2, 3}, 1, 3)));
    test(2, "bounded sum some in range", 19, () -> eq(9, Solution.boundedSum(new int[] {1, 2, 3, 4, 5}, 2, 4)));
    test(2, "bounded sum none in range", 20, () -> eq(0, Solution.boundedSum(new int[] {1, 2, 3}, 10, 20)));
    test(2, "bounded sum negative range", 21, () -> eq(2, Solution.boundedSum(new int[] {-5, -1, 0, 3}, -1, 3)));
    print();
  }
  static void test(int level, String name, int line, Fn fn) {
    if (level > maxLevel) { rows.add(new Row(name, "skip", null, line)); return; }
    try { fn.run(); rows.add(new Row(name, "pass", null, line)); }
    catch (Throwable e) { rows.add(new Row(name, "fail", e.toString(), line)); }
  }
  static void eq(Object expected, Object actual) {
    if (!Objects.equals(expected, actual)) throw new AssertionError("Expected " + expected + " but got " + actual);
  }
  static void print() {
    long p = rows.stream().filter(r -> r.status.equals("pass")).count();
    long f = rows.stream().filter(r -> r.status.equals("fail")).count();
    long s = rows.stream().filter(r -> r.status.equals("skip")).count();
    StringBuilder out = new StringBuilder("{\"rows\":[");
    for (int i = 0; i < rows.size(); i++) {
      Row r = rows.get(i);
      if (i > 0) out.append(",");
      out.append("{\"name\":\"").append(esc(r.name)).append("\",\"status\":\"").append(r.status).append("\",\"line\":").append(r.line);
      if (r.error != null) out.append(",\"error\":\"").append(esc(r.error)).append("\"");
      out.append("}");
    }
    out.append("],\"passed\":").append(p).append(",\"failed\":").append(f).append(",\"skipped\":").append(s).append("}");
    System.out.println(out);
  }
  static String esc(String s) { return s.replace("\\", "\\\\").replace("\"", "\\\"").replace("\n", "\\n"); }
}
